### ISSUE TITLE
Poor performance due to the default chunk size

### ISSUE THREAD

**Issue by gtca** [deshaw/versioned-hdf5#198]
There seems to be an issue with the default chunk size used. At least, that seems to be an issue coming from `h5py`.

Here's a simple snippet:

```python
import jax
import h5py
from versioned_hdf5 import VersionedHDF5File

z = np.array(jax.random.normal(jax.random.PRNGKey(1), shape=(10_000_000,)))
with h5py.File("zdemo.h5", "w") as f:
    f["z"] = z
os.stat("xydemo.h5").st_size / 1e6
# => 1.202048

file = h5py.File("zdemo_versioned.h5", "w")
vf = VersionedHDF5File(file)
with vf.stage_version("v1") as f:
    f["z"] = z
file.close()
vf.close()

os.stat("zdemo_versioned.h5").st_size / 1e6
# => 40.538272
```

It then gets more complicated to build interfaces on top of versioned-hdf5 when there are 1D and multi-dimensional arrays and neither a single chunk size can be provided, nor the defaults can be used:

https://github.com/deshaw/versioned-hdf5/blob/09f3a8f6cccdaa9d5eb28811139e4a327e519079/versioned_hdf5/wrappers.py#L193-L196

This leads to a question if it might be possible to rely on `h5py` [to choose an appropriate chunk size](https://docs.h5py.org/en/stable/high/dataset.html#chunked-storage) rather than to define it with the `DEFAULT_CHUNK_SIZE`. That should also work beyond 1D arrays then.

**Comment by asmeurer** [comment]
The default chunk size was chosen early on as something that appeared to have the best tradeoff between performance and file size, but we haven't investigated it in a while, so we might want to revisit it. 

The main issue is that due to the way Versioned HDF5 works, the chunk size has to remain constant across all versions of a dataset. That means that in general, you can't guess a good chunk size from the initial version of a dataset if its shape isn't representative of its typical shape across future versions. Also, the way it works is that dataset chunks are reused across versions, so the chunk size matters for more than just HDF5 read/write performance. It also directly affects how the dataset is stored in the versioned backend. In effect, all datasets work like they are set with `maxshape=(None, ..., None)`.

I think we could implement some sort of default chunk size for multidimensional datasets. It hasn't yet been something we've needed for our use-case, but it's something that would be convenient to do, although the caveats need to be documented.

**Comment by gtca** [comment]
Thanks for your elaborate answer, @asmeurer!

It makes sense, and default chunk sizes for mutlidimensional datasets would be great to have indeed.

I wonder if it might make sense to use or adopt [guess_chunk](https://github.com/h5py/h5py/blob/cff8537cc74a4897e2c0dc523309be17bec06dd4/h5py/_hl/filters.py#L331) from `h5py`.
That would mean that by default, the chunk size — across all the dataset versions — is to be chosen based on its initial version, which might be a decent default.

**Comment by peytondmurray** [comment]
I've been looking into implementing a call to `h5py._hl.filters.guess_chunks` [upon initial dataset creation](https://github.com/deshaw/versioned-hdf5/blob/a6d57df8527c7febb23b44685b2567b08c1b8665/versioned_hdf5/backend.py#L50). Initial testing indicates that this should work, but there are a few considerations here:

1. The chunks will be set when the dataset is first created. If it changes significantly in later revisions, it's possible the chunk size will not be suitable for future versions of a dataset. This has already been mentioned by @asmeurer above. With this in mind it might be worth writing a `rechunk` function to optimize the chunks across all versions.
2. We'd have to rewrite a bunch of the tests, some of which use `DEFAULT_CHUNK_SIZE` to perform certain checks against the structure of the data written to the file.

**Comment by peytondmurray** [comment]
Closed by #274.
